File: bt_service/core/control.py
"""Control flow nodes: Sequence, Fallback, ReactiveSequence, ReactiveFallback."""

from bt_service.core.node import Node, Haltable, Status


class Sequence(Node, Haltable):
    """Ticks children in order. Remembers position across ticks.

    FAILURE or RUNNING → return immediately, remember position.
    All children SUCCESS → SUCCESS.
    """

    def __init__(self, children=None):
        self.children = children or []
        self._index = 0

    def tick(self, bb) -> Status:
        while self._index < len(self.children):
            status = self.children[self._index].tick(bb)
            if status == Status.RUNNING:
                return Status.RUNNING
            elif status == Status.FAILURE:
                self._index = 0
                return Status.FAILURE
            elif status == Status.SUCCESS:
                self._index += 1
        self._index = 0
        return Status.SUCCESS

    def halt(self) -> None:
        if self._index < len(self.children):
            child = self.children[self._index]
            if isinstance(child, Haltable):
                child.halt()
        self._index = 0


class Fallback(Node, Haltable):
    """Tries each child until one succeeds. Remembers position.

    SUCCESS or RUNNING → return immediately, remember position.
    All children FAILURE → FAILURE.
    """

    def __init__(self, children=None):
        self.children = children or []
        self._index = 0

    def tick(self, bb) -> Status:
        while self._index < len(self.children):
            status = self.children[self._index].tick(bb)
            if status == Status.RUNNING:
                return Status.RUNNING
            elif status == Status.SUCCESS:
                self._index = 0
                return Status.SUCCESS
            elif status == Status.FAILURE:
                self._index += 1
        self._index = 0
        return Status.FAILURE

    def halt(self) -> None:
        if self._index < len(self.children):
            child = self.children[self._index]
            if isinstance(child, Haltable):
                child.halt()
        self._index = 0


class ReactiveSequence(Node, Haltable):
    """Ticks ALL children every tick. Resets non-RUNNING children each tick.

    First FAILURE → return FAILURE.
    All children SUCCESS → SUCCESS.
    """

    def __init__(self, children=None):
        self.children = children or []

    def tick(self, bb) -> Status:
        any_running = False
        for child in self.children:
            status = child.tick(bb)
            if status == Status.RUNNING:
                any_running = True
            elif status == Status.FAILURE:
                return Status.FAILURE
        return Status.RUNNING if any_running else Status.SUCCESS

    def halt(self) -> None:
        for child in self.children:
            if isinstance(child, Haltable):
                child.halt()


class ReactiveFallback(Node, Haltable):
    """Ticks ALL children every tick. First SUCCESS returns SUCCESS.

    All children FAILURE → FAILURE.
    """

    def __init__(self, children=None):
        self.children = children or []

    def tick(self, bb) -> Status:
        any_running = False
        all_failed = True
        for child in self.children:
            status = child.tick(bb)
            if status == Status.RUNNING:
                any_running = True
                all_failed = False
            elif status == Status.SUCCESS:
                return Status.SUCCESS
            elif status == Status.FAILURE:
                pass
        if any_running:
            return Status.RUNNING
        return Status.FAILURE

    def halt(self) -> None:
        for child in self.children:
            if isinstance(child, Haltable):
                child.halt()

File: pkg/bt/control.go
package bt

import "context"

// ---------------------------------------------------------------------------
// Sequence — ticks children in order.
//   FAILURE or RUNNING → return immediately, remember position.
//   All children SUCCESS → SUCCESS.
// ---------------------------------------------------------------------------

type Sequence struct {
	children []Node
	index    int
}

func NewSequence(children ...Node) *Sequence {
	return &Sequence{children: children}
}

func (s *Sequence) Tick(ctx context.Context) Status {
	for s.index < len(s.children) {
		status := s.children[s.index].Tick(ctx)
		switch status {
		case Running:
			return Running
		case Failure:
			s.index = 0
			return Failure
		case Success:
			s.index++
		}
	}
	// All children completed successfully
	s.index = 0
	return Success
}

func (s *Sequence) Halt() {
	if s.index < len(s.children) {
		if h, ok := s.children[s.index].(Haltable); ok {
			h.Halt()
		}
	}
	s.index = 0
}

// ---------------------------------------------------------------------------
// Fallback — tries each child until one succeeds.
//   SUCCESS or RUNNING → return immediately, remember position.
//   All children FAILURE → FAILURE.
// ---------------------------------------------------------------------------

type Fallback struct {
	children []Node
	index    int
}

func NewFallback(children ...Node) *Fallback {
	return &Fallback{children: children}
}

func (f *Fallback) Tick(ctx context.Context) Status {
	for f.index < len(f.children) {
		status := f.children[f.index].Tick(ctx)
		switch status {
		case Running:
			return Running
		case Success:
			f.index = 0
			return Success
		case Failure:
			f.index++
		}
	}
	f.index = 0
	return Failure
}

func (f *Fallback) Halt() {
	if f.index < len(f.children) {
		if h, ok := f.children[f.index].(Haltable); ok {
			h.Halt()
		}
	}
	f.index = 0
}

// ---------------------------------------------------------------------------
// ReactiveSequence — ticks ALL children every tick.
//   Resets non-RUNNING children each tick.
//   First FAILURE → return FAILURE.
//   All children SUCCESS → SUCCESS.
// ---------------------------------------------------------------------------

type ReactiveSequence struct {
	children []Node
}

func NewReactiveSequence(children ...Node) *ReactiveSequence {
	return &ReactiveSequence{children: children}
}

func (rs *ReactiveSequence) Tick(ctx context.Context) Status {
	anyRunning := false
	for _, child := range rs.children {
		status := child.Tick(ctx)
		switch status {
		case Running:
			anyRunning = true
		case Failure:
			return Failure
		}
	}
	if anyRunning {
		return Running
	}
	return Success
}

func (rs *ReactiveSequence) Halt() {
	for _, child := range rs.children {
		if h, ok := child.(Haltable); ok {
			h.Halt()
		}
	}
}

// ---------------------------------------------------------------------------
// ReactiveFallback — ticks ALL children every tick.
//   Resets non-RUNNING children each tick.
//   First child that is SUCCESS → SUCCESS.
//   All children FAILURE → FAILURE.
// ---------------------------------------------------------------------------

type ReactiveFallback struct {
	children []Node
}

func NewReactiveFallback(children ...Node) *ReactiveFallback {
	return &ReactiveFallback{children: children}
}

func (rf *ReactiveFallback) Tick(ctx context.Context) Status {
	anyRunning := false
	allFailed := true
	for _, child := range rf.children {
		status := child.Tick(ctx)
		switch status {
		case Running:
			anyRunning = true
			allFailed = false
		case Success:
			return Success
		case Failure:
			// continue checking remaining
		}
	}
	if anyRunning {
		return Running
	}
	if allFailed {
		return Failure
	}
	return Failure
}

func (rf *ReactiveFallback) Halt() {
	for _, child := range rf.children {
		if h, ok := child.(Haltable); ok {
			h.Halt()
		}
	}
}
